File: Samples/Mapping/Maps-Samples/Routing/PreRouting/PreRoutingViewController.m
//
//  PreRoutingViewController.m
//  PWMapKit
//
//  Created on 8/11/16.
//  Copyright © 2016 Phunware. All rights reserved.
//

#import <PureLayout/PureLayout.h>
#import <PWMapKit/PWMapKit.h>

#import "PreRoutingViewController.h"
#import "CommonSettings.h"
#import "RouteSummaryTableViewCell.h"
#import "PreRouteInstructionTableViewCell.h"
#import "RouteLabelTableViewCell.h"
#import "MapPreviewTableViewCell.h"
#import "ShowAllStepsTableViewCell.h"

static CGFloat const StartNavigationContainerViewHeight = 80.0;

@interface PreRoutingViewController () <UITableViewDelegate, UITableViewDataSource>

@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) UIView *startNavigationContainerView;
@property (nonatomic, assign) BOOL listViewIsSelected;
@property (nonatomic, strong) PWMapView *mapViewForPreview;
@property (nonatomic, strong) UISegmentedControl *segmentedControl;
@property (nonatomic, assign) BOOL showAllSteps;

@end

@implementation PreRoutingViewController

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    self.segmentedControl.hidden = YES;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setTitle:nil];
    self.listViewIsSelected = YES;
    self.showAllSteps = !UIAccessibilityIsVoiceOverRunning();
    
    [self configureTableView];
    [self configureStartNavigationContainerView];
    [self configureNavigationBar];
}

#pragma mark - Configuration

- (void)configureNavigationBar {
    NSString *listString = PWLocalizedString(@"SegmentTitleList", @"List");
    NSString *mapString = PWLocalizedString(@"SegmentTitleMap", @"Map");
    self.segmentedControl = [[UISegmentedControl alloc] initWithItems:@[listString, mapString]];
    self.segmentedControl.selectedSegmentIndex = 0;
    [self.segmentedControl addTarget:self action:@selector(segmentedControlChanged:) forControlEvents:UIControlEventValueChanged];
    [self.segmentedControl setWidth:80 forSegmentAtIndex:0];
    [self.segmentedControl setWidth:80 forSegmentAtIndex:1];
    
    [self.navigationController.navigationBar addSubview:self.segmentedControl];
    [self.segmentedControl autoCenterInSuperview];
}

- (void)configureTableView {
    self.tableView = [UITableView new];
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    self.tableView.contentInset = UIEdgeInsetsMake(0, 0, StartNavigationContainerViewHeight, 0);
    self.tableView.scrollIndicatorInsets = UIEdgeInsetsMake(0, 0, StartNavigationContainerViewHeight, 0);
    [self.tableView registerClass:[RouteSummaryTableViewCell class] forCellReuseIdentifier:RouteSummaryTableViewCellReuseIdentifier];
    [self.tableView registerClass:[PreRouteInstructionTableViewCell class] forCellReuseIdentifier:PreRouteInstructionCellReuseIdentifier];
    [self.tableView registerClass:[RouteLabelTableViewCell class] forCellReuseIdentifier:RouteLabelTableViewCellReuseIdentifier];
    [self.tableView registerClass:[MapPreviewTableViewCell class] forCellReuseIdentifier:MapPreviewTableViewCellReuseIdentifier];
    [self.tableView registerClass:[ShowAllStepsTableViewCell class] forCellReuseIdentifier:ShowAllStepsCellReuseIdentifier];
    [self.view addSubview:self.tableView];
    
    [self.tableView autoPinEdgesToSuperviewEdges];
}

- (void)configureStartNavigationContainerView {
    self.startNavigationContainerView = [UIView new];
    self.startNavigationContainerView.backgroundColor = [UIColor clearColor];
    [self.view addSubview:self.startNavigationContainerView];
    
    [self.startNavigationContainerView autoPinEdgeToSuperviewEdge:ALEdgeBottom];
    [self.startNavigationContainerView autoPinEdgeToSuperviewEdge:ALEdgeLeft];
    [self.startNavigationContainerView autoPinEdgeToSuperviewEdge:ALEdgeRight];
    [self.startNavigationContainerView autoSetDimension:ALDimensionHeight toSize:StartNavigationContainerViewHeight];
    
    UIButton *startNavigationButton = [UIButton new];
    startNavigationButton.backgroundColor = [CommonSettings commonButtonBackgroundColor];
    startNavigationButton.tintColor = [UIColor whiteColor];
    startNavigationButton.layer.cornerRadius = 25;
    startNavigationButton.clipsToBounds = YES;
    [startNavigationButton setTitle:PWLocalizedString(@"StartNavigationButtonTitle", @"START NAVIGATION") forState:UIControlStateNormal];
    startNavigationButton.titleLabel.font = [UIFont boldSystemFontOfSize:20.0];
    [startNavigationButton addTarget:self action:@selector(startNavigationButtonTapped) forControlEvents:UIControlEventTouchUpInside];
    [self.startNavigationContainerView addSubview:startNavigationButton];
    
    [startNavigationButton autoAlignAxisToSuperviewAxis:ALAxisHorizontal];
    [startNavigationButton autoAlignAxisToSuperviewAxis:ALAxisVertical];
    [startNavigationButton autoMatchDimension:ALDimensionHeight toDimension:ALDimensionHeight ofView:self.startNavigationContainerView withMultiplier:0.6];
    [startNavigationButton autoMatchDimension:ALDimensionWidth toDimension:ALDimensionWidth ofView:self.startNavigationContainerView withMultiplier:0.9];
}

#pragma mark - Table View DataSource and Delegate

-(CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 44;
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    CGFloat height = UITableViewAutomaticDimension;
    if (!self.listViewIsSelected && indexPath.row == 1) {
        height = self.tableView.contentSize.height;
    }
    return height;
}

- (NSInteger) numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSInteger numberOfRows = self.route.routeInstructions.count + 2;
    if (!self.listViewIsSelected) {
        numberOfRows = 2;
    } else if (!self.showAllSteps) {
        return 2;
    }
    return numberOfRows;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = nil;
    if (indexPath.row == 0) {
        RouteSummaryTableViewCell *routeSummaryTableViewCell = [tableView dequeueReusableCellWithIdentifier:RouteSummaryTableViewCellReuseIdentifier];
        [routeSummaryTableViewCell configureForRoute:self.route];
        
        cell = routeSummaryTableViewCell;
    } else {
        if (self.listViewIsSelected) {
            cell = [self cellForListViewAtIndexPath:indexPath tableView:tableView];
        } else {
            cell = [self cellForMapViewAtIndexPath:indexPath tableView:tableView];
        }
    }
    
    return cell;
}

- (UITableViewCell *)cellForListViewAtIndexPath:(NSIndexPath *)indexPath tableView:(UITableView *)tableView {
    UITableViewCell *cell = nil;
    if (indexPath.row == 1 && self.showAllSteps) {
        RouteLabelTableViewCell *routeLabelTableViewCell = [tableView dequeueReusableCellWithIdentifier:RouteLabelTableViewCellReuseIdentifier];
        [routeLabelTableViewCell configure];
        cell = routeLabelTableViewCell;
    } else if (self.showAllSteps) {
        PreRouteInstructionTableViewCell *preRouteInstructionTableViewCell = [tableView dequeueReusableCellWithIdentifier:PreRouteInstructionCellReuseIdentifier];
        
        PWRouteInstruction *routeInstruction = self.route.routeInstructions[indexPath.row - 2];
        [preRouteInstructionTableViewCell configureForRouteInstruction:routeInstruction];
        
        cell = preRouteInstructionTableViewCell;
    } else {
        ShowAllStepsTableViewCell *showAllStepsCell = [tableView dequeueReusableCellWithIdentifier:ShowAllStepsCellReuseIdentifier];
        
        [showAllStepsCell configureForShowAllStepsSelector:@selector(showAllStepsButtonTapped) showAllStepsTarget:self];
        
        cell = showAllStepsCell;
    }
    return cell;
}

- (UITableViewCell *)cellForMapViewAtIndexPath:(NSIndexPath *)indexPath tableView:(UITableView *)tableView {
    UITableViewCell *cell = nil;
    if (indexPath.row == 1) {
        if (self.mapViewForPreview == nil) {
            self.mapViewForPreview = [PWMapView new];
            [self.mapViewForPreview setBuilding:self.route.building];
        }
        MapPreviewTableViewCell *mapPreviewCell = [tableView dequeueReusableCellWithIdentifier:MapPreviewTableViewCellReuseIdentifier];
        [mapPreviewCell configureForRoute:self.route mapView:self.mapViewForPreview];
        cell = mapPreviewCell;
    }
    return cell;
}

#pragma mark - UI Actions

- (void)startNavigationButtonTapped {
    [self.segmentedControl setHidden:YES];
    [self.navigationController popToRootViewControllerAnimated:YES];
    
    // Notify to plot the route
    [[NSNotificationCenter defaultCenter] postNotificationName:PlotRouteNotification object:self.route];
}

- (void)backButtonTapped {
    [self.navigationController popViewControllerAnimated:YES];
}

- (void)segmentedControlChanged:(id)sender {
    UISegmentedControl *segmentedControl = sender;
    
    if (segmentedControl.selectedSegmentIndex == 0) {
        [self listSegmentSelected];
    } else {
        [self mapSegmentSelected];
    }
}

- (void)listSegmentSelected {
    self.listViewIsSelected = YES;
    [self.tableView reloadData];
}

- (void)mapSegmentSelected {
    self.listViewIsSelected = NO;
    [self.tableView reloadData];
}

- (void)showAllStepsButtonTapped {
    self.showAllSteps = YES;
    [self.tableView reloadData];
}

@end

File: Samples/PhunMaps/PhunMaps/Controllers/PreRoutingViewController.swift
//
//  PreRoutingViewController.swift
//  Mapping-Sample
//
//  Created on 6/16/17.
//  Copyright © 2017 Phunware. All rights reserved.
//

import UIKit
import PWMapKit

class PreRoutingViewController: UIViewController {
    
    enum rowMap: Int {
        case summary
        case routeLabel
        case showAllSteps
        case instruction
        case mapPreview
    }
    
    enum segment: String {
        case list = "List"
        case map = "Map"
        
        func localizedString() -> String {
            return NSLocalizedString(self.rawValue, comment: "")
        }
    }
    
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var startNavigationButton: UIButton!
    @IBOutlet weak var startNavigationButtonBottomConstraint: NSLayoutConstraint!
    
    var segmentedControl: UISegmentedControl!
    let segmentsList: [segment] = [.list, .map]
    var route: PWRoute?
    var showAllInstructions = !UIAccessibilityIsVoiceOverRunning()
    let mapViewForPreview = PWMapView()
    
    fileprivate var rowMapping = [rowMap]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.rowHeight = UITableViewAutomaticDimension
        tableView.estimatedRowHeight = 44.0
        tableView.contentInset = UIEdgeInsetsMake(0.0, 0.0, startNavigationButton.frame.height+startNavigationButtonBottomConstraint.constant, 0.0)
        
        startNavigationButton.isHidden = route == nil
        startNavigationButton.layer.cornerRadius = 25
        startNavigationButton.clipsToBounds = true
        startNavigationButton.backgroundColor = CommonSettings.buttonBackgroundColor
        startNavigationButton.setTitle(NSLocalizedString("START NAVIGATION", comment: ""), for: .normal)
        startNavigationButton.addTarget(self, action: #selector(startNavigation(sender:)), for: .touchUpInside)
        
        if let route = route {
            mapViewForPreview.setBuilding(route.building)
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        configureRowMapping(selectedSegment: .list)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        configureSegmentedControl()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        segmentedControl.isHidden = true
        navigationItem.titleView = nil
        super.viewWillDisappear(animated)
    }
    
    func configureRowMapping(selectedSegment: segment) {
        rowMapping = [rowMap]()
        rowMapping.append(.summary)
        
        switch selectedSegment {
        case .list:
            if showAllInstructions {
                rowMapping.append(.routeLabel)
                if let route = route {
                    for _ in 0..<route.routeInstructions.count {
                        rowMapping.append(.instruction)
                    }
                }
            } else {
                rowMapping.append(.showAllSteps)
            }
        case .map:
            rowMapping.append(.mapPreview)
        }
    }
}

// MARK: - Segmented Control

extension PreRoutingViewController {
    
    func configureSegmentedControl() {
        segmentedControl = UISegmentedControl()
        navigationItem.titleView = segmentedControl
        segmentedControl.heightAnchor.constraint(equalToConstant: 30.0).isActive = true
        segmentedControl.translatesAutoresizingMaskIntoConstraints = false
        
        for index in 0..<segmentsList.count {
            segmentedControl.insertSegment(withTitle: segmentsList[index].localizedString(), at: index, animated: false)
            segmentedControl.setWidth(80.0, forSegmentAt: index)
        }
        
        segmentedControl.addTarget(self, action: #selector(segmentedControlChanged(_:)), for: .valueChanged)
        segmentedControl.selectedSegmentIndex = 0
    }
    
    func segmentedControlChanged(_ sender: UISegmentedControl) {
        configureRowMapping(selectedSegment: segmentsList[segmentedControl.selectedSegmentIndex])
        tableView.reloadData()
    }
}

// MARK: - UITableViewDelegate

extension PreRoutingViewController: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
        return 44.0
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        var height = UITableViewAutomaticDimension
        if rowMapping[indexPath.row] == .mapPreview {
            height = 400.0
        }
        return height
    }
}

// MARK: - UITableViewDataSource

extension PreRoutingViewController: UITableViewDataSource {

    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return rowMapping.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell = UITableViewCell()
        
        switch rowMapping[indexPath.row] {
        case .summary:
            if let routeSummaryCell = tableView.dequeueReusableCell(withIdentifier: String(describing: RouteSummaryCell.self), for: indexPath) as? RouteSummaryCell {
                if let route = route {
                    routeSummaryCell.configureForRoute(route)
                }
                cell = routeSummaryCell
            }
        case .routeLabel:
            if let routeLabelCell = tableView.dequeueReusableCell(withIdentifier: String(describing: RouteLabelCell.self), for: indexPath) as? RouteLabelCell {
                cell = routeLabelCell
            }
        case .showAllSteps:
            if let showAllInstructionsCell = tableView.dequeueReusableCell(withIdentifier: String(describing: ShowAllInstructionsCell.self), for: indexPath) as? ShowAllInstructionsCell {
                showAllInstructionsCell.configureForShowStepsSelector(#selector(showAllInstructions(sender:)), self)
                cell = showAllInstructionsCell
            }
        case .instruction:
            if let instructionCell = tableView.dequeueReusableCell(withIdentifier: String(describing: PreRouteInstructionCell.self), for: indexPath) as? PreRouteInstructionCell {
                if let indexOffset = rowMapping.index(of: .instruction), let instruction = route?.routeInstructions[indexPath.row-indexOffset]as? PWRouteInstruction {
                    instructionCell.configure(forInstruction: instruction)
                }
                cell = instructionCell
            }
        case .mapPreview:
            if let mapPreviewCell = tableView.dequeueReusableCell(withIdentifier: String(describing: MapPreviewCell.self), for: indexPath) as? MapPreviewCell {
                if let route = route {
                    mapPreviewCell.configureFor(mapView: mapViewForPreview, route: route)
                }
                cell = mapPreviewCell
            }
        }
        
        return cell
    }
}

// MARK: - Routing Actions

extension PreRoutingViewController {
    
    func showAllInstructions(sender: UIButton) {
        showAllInstructions = true
        configureRowMapping(selectedSegment: segmentsList[segmentedControl.selectedSegmentIndex])
        tableView.reloadData()
    }
    
    func startNavigation(sender: UIButton) {
        segmentedControl.isHidden = true
        navigationController?.popToRootViewController(animated: true)
        
        if let route = route {
            NotificationCenter.default.post(name: .startNavigatingRoute, object: route)
        }
    }
}
